### ISSUE TITLE
Security warning: avoid using function constructor.

### ISSUE THREAD

**Issue by feugy** [plotly/plotly.js#897]
I recently discovered that one of plotly.js's dependency is using `Function` constructor.
Unfortunately, I can't tell exactly which one, but the code doesn't seems to belong to plotly itself.

Here is Chrome's error message:
```
plotly.js:74222Uncaught EvalError: Refused to evaluate a string as JavaScript because 'unsafe-eval' is
 not an allowed source of script in the following Content Security Policy directive: "script-src 'self'
 www.google-analytics.com ajax.googleapis.com 'nonce-67df8dace6ea7feb57cab73e41ebe0c7'".

makeOp	@	plotly.js:74222
(anonymous function)	@	plotly.js:74244
455.cwise-compiler	@	plotly.js:74271
s	@	plotly.js:7
(anonymous function)	@	plotly.js:7
129.ndarray	@	plotly.js:25687
s	@	plotly.js:7
(anonymous function)	@	plotly.js:7
138../lib/shaders	@	plotly.js:27330
s	@	plotly.js:7
(anonymous function)	@	plotly.js:7
798.../../constants/gl3d_dashes	@	plotly.js:127609
s	@	plotly.js:7
(anonymous function)	@	plotly.js:7
800.../../constants/gl_markers	@	plotly.js:128154
s	@	plotly.js:7
(anonymous function)	@	plotly.js:7
15.../src/traces/scatter3d	@	plotly.js:355
s	@	plotly.js:7
(anonymous function)	@	plotly.js:7
12../bar	@	plotly.js:312
s	@	plotly.js:7
e	@	plotly.js:7
(anonymous function)	@	plotly.js:7
a	@	plotly.js:7
(anonymous function)	@	plotly.js:7
__webpack_require__	@	bootstrap 08e5224…:585
fn	@	bootstrap 08e5224…:109
(anonymous function)	@	svg-to-png.js:2
__webpack_require__	@	bootstrap 08e5224…:585
fn	@	bootstrap 08e5224…:109
(anonymous function)	@	commons.js:36658
__webpack_require__	@	bootstrap 08e5224…:585
fn	@	bootstrap 08e5224…:109
(anonymous function)	@	index.js:1
__webpack_require__	@	bootstrap 08e5224…:585
fn	@	bootstrap 08e5224…:109
(anonymous function)	@	bootstrap.js:9
__webpack_require__	@	bootstrap 08e5224…:585
fn	@	bootstrap 08e5224…:109
(anonymous function)	@	index.js:3
__webpack_require__	@	bootstrap 08e5224…:585
fn	@	bootstrap 08e5224…:109
(anonymous function)	@	index.js:8
__webpack_require__	@	bootstrap 08e5224…:585
webpackJsonpCallback	@	bootstrap 08e5224…:21
(anonymous function)	@	index.js:1
```

Function constructor is considered by browser's CSP as an equivalent of `eval`.
Therefore, we can't setup a good Content Security Policy, unless we allow `'unsafe-eval'` which is a whole in the policy itself.

It is possible to identify the faulty dependency, and maybe provide a workaround ?

**Comment by rreusser** [comment]
I think this is the line: https://github.com/scijs/cwise-compiler/blob/master/lib/compile.js#L351

cwise is a library that performs operations on ndarrays. It doesn't look like plotly uses the [cwise browserify transform](https://github.com/scijs/cwise#as-a-browserify-transform) when [bundling plotly.js](https://github.com/plotly/plotly.js/blob/master/tasks/bundle.js#L72), but I believe even the transformed version that does the parsing ahead of time still compiles an optimized version (i.e. `new Function`) at runtime.

Long story short, this is pretty central to what cwise does, so I'm not totally sure of a good workaround, unless there's a way to short-circuit the optimization and just perform naive operations on data.

**Comment by etpinard** [comment]
> It doesn't look like plotly uses the cwise browserify transform when bundling plotly.js,

The plotly build step doesn't explicitly list the `cwise` transform, **but** several of our dependencies have `cwise` listed in their respective package.json namely `ndarray-fill`, `ndarray-wrap` and `gl-select-static` (ref this [script](https://gist.github.com/etpinard/b11ee1ae958abfc50be5)).

**Comment by willemx** [comment]
I'm also running into this problem. Can't run my (Meteor) app because of this.
Is there a work-around maybe?

**Comment by jacobq** [comment]
When I tried to bring plotly.js into my Electron application (via [ember-plotly-shim](https://github.com/brianhjelle/ember-plotly-shim)) it triggers a security warning. In fact, since I have a set a Content-Security-Policy [per the recommendations](https://electronjs.org/docs/tutorial/security#define-a-content-security-policy), it does not run at all. 

Would someone be willing to write some / point to some documentation that describes how to incorporate plotly w/o using a pipeline that evaluates strings as code?

![unsafe-string-eval](https://user-images.githubusercontent.com/1633459/36735475-ee2059c4-1b9b-11e8-80c1-cfb92fefb5cc.png)

**Comment by fxfilmxf** [comment]
Still seeing this as well. Hasn't been addressed at all.

**Comment by etpinard** [comment]
Nothing has changed since https://github.com/plotly/plotly.js/issues/897#issuecomment-244402532, so yes this hasn't been addressed _at all_ :scream: 

The problematic `cwise` transform is only used in gl3d and gl2d (excluding `scattergl`), so using one of the [partial bundles](https://github.com/plotly/plotly.js/tree/master/dist#partial-bundles) that don't include these trace types would _solve_ this issue. Better yet, bundling plotly.js [yourself](https://github.com/plotly/plotly.js#modules) may do the trick.

**Comment by fxfilmxf** [comment]
Thanks for the advice! I'll take a look at that.

**Comment by jtal** [comment]
Does anyone have the CSP header they've used to keep security as tight as possible but still use plotly?

**Comment by jacobq** [comment]
@jtal 

> Does anyone have the CSP header they've used to keep security as tight as possible but still use plotly?

@jtal I recommend starting with the strictest options and gradually relaxing restrictions until your application functions properly. There may be other parts of your work that won't work when fully restricted. As an example only, here is one I have used (inside of an electron app):
`Content-Security-Policy: default-src 'none'; script-src 'self' 'unsafe-inline' 'unsafe-eval'; connect-src 'self'; img-src 'self' 'data:'; style-src 'self' 'unsafe-inline'`

See [MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP)

**Comment by anders-kiaer** [comment]
Looks like the direct `plotly.js` dependencies today that make `plotly.js` dependent on [`cwise`](https://github.com/scijs/cwise) are:
 - [`ndarray-fill`](https://github.com/scijs/ndarray-fill)
 - [`ndarray-homography`](https://github.com/scijs/ndarray-homography)
 - [`gl-plot2d`](https://github.com/gl-vis/gl-plot2d)
 - [`gl-plot3d`](https://github.com/gl-vis/gl-plot3d)

Does anyone know if it is feasible for `plotly.js` to go away from these dependencies (any alternatives?) and/or modify the dependencies themselves to not use `cwise`?

<sup>The [`unsafe-eval`](https://developer.chrome.com/extensions/contentSecurityPolicy#relaxing-eval) requirement is unfortunately currently stopping us from being allowed to use `plotly.js`'s variant of [parallell coordinates](https://plot.ly/javascript/parallel-coordinates-plot/), as it appears to be part of the [`gl2d` bundle](https://github.com/plotly/plotly.js/blob/master/dist/README.md#plotlyjs-gl2d), which then looks to be indirectly requiring `cwise`, ref. https://github.com/plotly/plotly.js/issues/897#issuecomment-371968332.</sup>

**Comment by etpinard** [comment]
@anders-kiaer you may be interested in https://www.npmjs.com/package/plotly.js-gl2d-dist, you can install it using

```
npm install plotly.js-gl2d-dist
```

which has no dependencies.

**Comment by anders-kiaer** [comment]
Thanks @etpinard 🙂  Hmm, not sure if I understand, isn't the `cwise` dependency bundled into the `gl2d` dist? I tried using the dist from `npm install plotly.js-gl2d-dist`, and also what I think is the CDN equivalent, and got the same CSP block without `'unsafe-eval'`. Tested using this Dash snippet:
```python
import dash
import dash_html_components as html
from flask_talisman import Talisman

app = dash.Dash(
    __name__, external_scripts=["https://cdn.plot.ly/plotly-gl2d-latest.min.js"]
)


CSP = {
    "default-src": "'self'",
    "script-src": [
        "'self'",
        "https://cdn.plot.ly",
        # "'unsafe-eval'",
        "'sha256-jZlsGVOhUAIcH+4PVs7QuGZkthRMgvT2n0ilH6/zTM0='",
    ],
    "style-src": ["'self'", "'unsafe-inline'"],
}

Talisman(app.server, content_security_policy=CSP, force_https=False)

app.layout = html.Span("Hello Dash!")

if __name__ == "__main__":
    app.run_server()
```
which when looking at the browser console gives a
```
EvalError: call to Function() blocked by CSP
```
unless `'unsafe-eval'` is added. Searching through [`plotly-gl2d.js`](https://raw.githubusercontent.com/plotly/plotly.js/master/dist/plotly-gl2d.js) gives me 18 hits on ` Function(`. The first hit looks to come from [this line](https://github.com/mikolalysenko/binary-search-bounds/blob/e5bc6aeb8fb81bfade3db977f36c115f17445afc/search-bounds.js#L32).

**Comment by etpinard** [comment]
My mistake. I thought you were worried about the "unsafe-eval" behavior of `cwise` when bundling your own custom bundle, not in the browser runtime. 

Unfortunately, there's no easy way for us to simply replace `cwise`. This package can improve performance by a large amount.

**Comment by etpinard** [comment]
But, the `basic`, `cartesian`, `geo`, `mapbox` and `finance` [partial bundles](https://github.com/plotly/plotly.js/tree/master/dist#partial-bundles) should be fine.

That is, bundles that include trace types other than gl3d, gl2d, scattergl and splom should pass the "unsafe-eval" checks.

**Comment by alexcjohnson** [comment]
FWIW I don't think it would be unreasonable to explore making an eval-free version of `cwise` - it's not the `eval` that gives it its speed, it's just using that to do codegen and reduce its own size. But at the end of the day I'd expect gzip means this codegen isn't even saving that many bytes over the wire... not a guarantee, but if someone wants to explore that we would certainly entertain it.

**Comment by tdelmas** [comment]
The current version 1.52.2 (https://raw.githubusercontent.com/plotly/plotly.js/v1.52.2/dist/plotly.js) uses `Fonction` in eval mode in those places:
- [ ] L6793 In `compileBoundsSearch` from https://github.com/mikolalysenko/binary-search-bounds can be fixed by https://github.com/mikolalysenko/binary-search-bounds/pull/8 (Using a fork of this package will not be useful as the package using it is from the same author)
- [ ] L10908 In `bruteForcePlanner` from https://github.com/mikolalysenko/box-intersect (an issue is opened: https://github.com/mikolalysenko/box-intersect/issues/6) required by clean-pslg 
- [ ] L15277 in `compileBoundsSearch`again (!) (probably two different version of binary-search-bounds required by two different packages...)
- [ ] L18041 in `generateCWiseOp`
- [ ] L18128 in `createThunk`
- [ ] L26997 in `dsv.parseRows`
- [x] L37883 in `polyfill` but that line should never be reached in browser context so it's fine.
- [ ] L50960 in `identity`
- [ ] L50975 in `makeGetter`
- [ ] L51056 in `makeSetter`
- [ ] L57863 in `bakeOrient`
- [ ] L60170 in `compileSurfaceProcedure`
- [ ] L60729 in `pcompile`
- [ ] L61362 / L61365 in `createInsertionSort`
- [ ] L61816 / L61819 in `createQuickSort`
- [ ] L61892 in `compileSort`
- [ ] L61998 / L62031 / L62202 in `compileConstructor`
- [ ] L69845 in `compileDeterminant`
- [ ] L70016 in `orientation`
- [ ] L70098 in `generateSolver`
- [ ] L70225 in `orientation` (bis)
- [ ] L71041 in `createCellPolygonizer`
- [ ] L72727 in `buildSurfaceNets`

**Comment by archmoj** [comment]
> Looks like the direct `plotly.js` dependencies today that make `plotly.js` dependent on [`cwise`](https://github.com/scijs/cwise) are:
> 
> * [`ndarray-fill`](https://github.com/scijs/ndarray-fill)
> * [`ndarray-homography`](https://github.com/scijs/ndarray-homography)
> * [`gl-plot2d`](https://github.com/gl-vis/gl-plot2d)
> * [`gl-plot3d`](https://github.com/gl-vis/gl-plot3d)
> 
> Does anyone know if it is feasible for `plotly.js` to go away from these dependencies (any alternatives?) and/or modify the dependencies themselves to not use `cwise`?


@anders-kiaer this is done in #4929 and #4930 and available in plotly.js >= v1.54.4

**Comment by polx** [comment]
We're still meeting the use of eval as a problem trying to integrate plotly.js within CryptPad.
Using smaller ones as anders suggested (in https://github.com/plotly/plotly.js/issues/897#issuecomment-585315283 did not solve it for us).

A pity...

**Comment by nicolaskruchten** [comment]
@polx can you confirm this is happening with version 1.54.7 of plotly.js, and can you provide some context around what tool you're using that is complaining about this please? This issue has wandered all over the place since 2016 and we recently addressed parts of it but it's not clear what's outstanding. It might be helpful to create a new issue narrowly-scope to the problem you're seeing.

**Comment by polx** [comment]
Hello @nicolaskruchten ,
Yes, it was on 1.54.7. We've been trying to embed plotly renderings as one of the rendering "macros" for cryptpad code (https://cryptpad.fr/). Cryptpad aimst at very high security standard and thus requires CSP. 
If we deactivate CSP, then things load.
If we activate CSP then I get the following having replaced the plotly with the non-minimized version.

        Uncaught EvalError: call to Function() blocked by CSP
        makeOp http://localhost:3000/lib/plotly-latest.min.js?ver=3.20.1-1596722563915:60786
        [443]</< http://localhost:3000/lib/plotly-latest.min.js?ver=3.20.1-1596722563915:60808
        [443]< http://localhost:3000/lib/plotly-latest.min.js?ver=3.20.1-1596722563915:60835
        o http://localhost:3000/lib/plotly-latest.min.js?ver=3.20.1-1596722563915:7
        o http://localhost:3000/lib/plotly-latest.min.js?ver=3.20.1-1596722563915:7
        [253]< http://localhost:3000/lib/plotly-latest.min.js?ver=3.20.1-1596722563915:41713
        o http://localhost:3000/lib/plotly-latest.min.js?ver=3.20.1-1596722563915:7
        o http://localhost:3000/lib/plotly-latest.min.js?ver=3.20.1-1596722563915:7
        [266]< http://localhost:3000/lib/plotly-latest.min.js?ver=3.20.1-1596722563915:44060
        o http://localhost:3000/lib/plotly-latest.min.js?ver=3.20.1-1596722563915:7
        o http://localhost:3000/lib/plotly-latest.min.js?ver=3.20.1-1596722563915:7
        [1164]< http://localhost:3000/lib/plotly-latest.min.js?ver=3.20.1-1596722563915:190058
        o http://localhost:3000/lib/plotly-latest.min.js?ver=3.20.1-1596722563915:7
        o http://localhost:3000/lib/plotly-latest.min.js?ver=3.20.1-1596722563915:7
        [1166]< http://localhost:3000/lib/plotly-latest.min.js?ver=3.20.1-1596722563915:190689
        o http://localhost:3000/lib/plotly-latest.min.js?ver=3.20.1-1596722563915:7
        o http://localhost:3000/lib/plotly-latest.min.js?ver=3.20.1-1596722563915:7
        [36]< http://localhost:3000/lib/plotly-latest.min.js?ver=3.20.1-1596722563915:614
        o http://localhost:3000/lib/plotly-latest.min.js?ver=3.20.1-1596722563915:7
        o http://localhost:3000/lib/plotly-latest.min.js?ver=3.20.1-1596722563915:7
        [26]< http://localhost:3000/lib/plotly-latest.min.js?ver=3.20.1-1596722563915:420
        o http://localhost:3000/lib/plotly-latest.min.js?ver=3.20.1-1596722563915:7
        r http://localhost:3000/lib/plotly-latest.min.js?ver=3.20.1-1596722563915:7
        <anonymous> http://localhost:3000/lib/plotly-latest.min.js?ver=3.20.1-1596722563915:7
        execCb http://localhost:3001/bower_components/requirejs/require.js?ver=2.3.5:1696

Does it help?

Paul

**Comment by urnansbald** [comment]
> @anders-kiaer this is done in #4929 and #4930 and available in plotly.js >= v1.54.4

I'm also still experiencing this eval error with v1.54.7 @nicolaskruchten 
both pages are HTML and one is using the CDN and the other one has v1.54.7 in a <script>
CSP: <name="Content-Security-Policy" value="default-src 'self' 'unsafe-inline' *.bootstrapcdn.com" />

`Uncaught EvalError: call to Function() blocked by CSP
    o https://cdn.plot.ly/plotly-latest.min.js:35
    [443]</< https://cdn.plot.ly/plotly-latest.min.js:35
    [443]< https://cdn.plot.ly/plotly-latest.min.js:35
    a https://cdn.plot.ly/plotly-latest.min.js:7
     ...
Content Security Policy: The page’s settings blocked the loading of a resource at eval (“default-src”). plotly-latest.min.js:35:783306
Uncaught EvalError: call to Function() blocked by CSP
    o https://xxx.azurewebsites.net/Graph.html:42
    [443]</< https://xxx.azurewebsites.net/Graph.html:42
    [443]< https://xxx.azurewebsites.net/Graph.html:42
    a https://xxx.azurewebsites.net/Graph.html:14
    ...`

**Comment by polx** [comment]
For cryptpad, we have made a branch public https://github.com/xwiki-labs/cryptpad/tree/plotly_hackathon .
There is a setting to remove CPS but we'd like not to share it publicly. Please ask us offline how to do that (e.g. using my contacts page).

**Comment by archmoj** [comment]
> Hello @nicolaskruchten ,
> Yes, it was on 1.54.7. We've been trying to embed plotly renderings as one of the rendering "macros" for cryptpad code (https://cryptpad.fr/). Cryptpad aimst at very high security standard and thus requires CSP.
> If we deactivate CSP, then things load.
> If we activate CSP then I get the following having replaced the plotly with the non-minimized version.
> 
> ```
>     Uncaught EvalError: call to Function() blocked by CSP
>     makeOp http://localhost:3000/lib/plotly-latest.min.js?ver=3.20.1-1596722563915:60786
>     [443]</< http://localhost:3000/lib/plotly-latest.min.js?ver=3.20.1-1596722563915:60808
>     [443]< http://localhost:3000/lib/plotly-latest.min.js?ver=3.20.1-1596722563915:60835
>     o http://localhost:3000/lib/plotly-latest.min.js?ver=3.20.1-1596722563915:7
>     o http://localhost:3000/lib/plotly-latest.min.js?ver=3.20.1-1596722563915:7
>     [253]< http://localhost:3000/lib/plotly-latest.min.js?ver=3.20.1-1596722563915:41713
>     o http://localhost:3000/lib/plotly-latest.min.js?ver=3.20.1-1596722563915:7
>     o http://localhost:3000/lib/plotly-latest.min.js?ver=3.20.1-1596722563915:7
>     [266]< http://localhost:3000/lib/plotly-latest.min.js?ver=3.20.1-1596722563915:44060
>     o http://localhost:3000/lib/plotly-latest.min.js?ver=3.20.1-1596722563915:7
>     o http://localhost:3000/lib/plotly-latest.min.js?ver=3.20.1-1596722563915:7
>     [1164]< http://localhost:3000/lib/plotly-latest.min.js?ver=3.20.1-1596722563915:190058
>     o http://localhost:3000/lib/plotly-latest.min.js?ver=3.20.1-1596722563915:7
>     o http://localhost:3000/lib/plotly-latest.min.js?ver=3.20.1-1596722563915:7
>     [1166]< http://localhost:3000/lib/plotly-latest.min.js?ver=3.20.1-1596722563915:190689
>     o http://localhost:3000/lib/plotly-latest.min.js?ver=3.20.1-1596722563915:7
>     o http://localhost:3000/lib/plotly-latest.min.js?ver=3.20.1-1596722563915:7
>     [36]< http://localhost:3000/lib/plotly-latest.min.js?ver=3.20.1-1596722563915:614
>     o http://localhost:3000/lib/plotly-latest.min.js?ver=3.20.1-1596722563915:7
>     o http://localhost:3000/lib/plotly-latest.min.js?ver=3.20.1-1596722563915:7
>     [26]< http://localhost:3000/lib/plotly-latest.min.js?ver=3.20.1-1596722563915:420
>     o http://localhost:3000/lib/plotly-latest.min.js?ver=3.20.1-1596722563915:7
>     r http://localhost:3000/lib/plotly-latest.min.js?ver=3.20.1-1596722563915:7
>     <anonymous> http://localhost:3000/lib/plotly-latest.min.js?ver=3.20.1-1596722563915:7
>     execCb http://localhost:3001/bower_components/requirejs/require.js?ver=2.3.5:1696
> ```
> 
> Does it help?
> 
> Paul

@polx could you please use the unminified version i.e. "plotly-latest.js" instead to re-produce the list?

**Comment by polx** [comment]
Sorry @archmoj , it is called plotly-min but is a copy of plotly.js.
I just downloaded 1.54.7 again and compared and they are equal:
`diff lib/plotly-latest.min.js /tmp/plotly.js-1.54.7/dist/plotly.js` is empty (and both have md5 8948b85ca08520fa7ff3718ffe40dd4e).

paul

**Comment by jackparmer** [comment]
This issue has been tagged with `NEEDS SPON$OR`

A [community PR](https://github.com/plotly/plotly.js/pulls?q=is%3Apr+is%3Aopen+label%3A%22type%3A+community%22) for this feature would certainly be welcome, but our experience is deeper features like this are difficult to complete without the Plotly maintainers leading the effort.

**Sponsorship range: $45k-$50k**

What Sponsorship includes:
- Completion of this feature to the Sponsor's satisfaction, in a manner coherent with the rest of the Plotly.js library and API
- Tests for this feature
- Long-term support (continued support of this feature in the latest version of Plotly.js)
- Documentation at [plotly.com/javascript](https://plotly.com/javascript/)
- Possibility of integrating this feature with Plotly Graphing Libraries (Python, R, F#, Julia, MATLAB, etc)
- Possibility of integrating this feature with Dash
- Feature announcement on [community.plotly.com](https://community.plotly.com/tag/announcements) with shout out to Sponsor (or can remain anonymous)
- Gratification of advancing the world's most downloaded, interactive scientific graphing libraries (>50M downloads across supported languages)

Please include the link to this issue when [contacting us](https://plotly.com/get-pricing/) to discuss.

**Comment by nicolaskruchten** [comment]
Great news: [Equinor](https://www.equinor.com/en.html) is generously willing to sponsor $15k of this work if we can find sponsors for the rest! Please let us know if you're able to contribute part of the remaining $30k-$35k and we can get started on this project :)

**Comment by archmoj** [comment]
Here is the complete list of function constructors and their dependency path from different bundles.
# dependency: d3
```
plotly.js@1.58.1
└── d3@3.5.17
```

## at:
    file: d3/src/dsv/dsv.js
    func: dsv.parse
    line: var a = new Function("d", "return {" + row.map(function(name, i) {

| bundle    | impacted |
| --------- | -------- |
| basic     | yes      |
| cartesian | yes      |
| finance   | yes      |
| geo       | yes      |
| mapbox    | yes      |
| gl2d      | yes      |
| gl3d      | yes      |
---

# dependency: es6-promise
```
plotly.js@1.58.1
└── es6-promise@4.2.8
```

## at:
    file: es6-promise/dist/es6-promise.js
    func: attemptVertx
    line: var vertx = Function('return this')().require('vertx');

and
## at:
    file: es6-promise/dist/es6-promise.js
    func: polyfill
    line: local = Function('return this')();

| bundle    | impacted |
| --------- | -------- |
| basic     | yes      |
| cartesian | yes      |
| finance   | yes      |
| geo       | yes      |
| mapbox    | yes      |
| gl2d      | yes      |
| gl3d      | yes      |
---

# dependency: binary-search-bounds
```
plotly.js@1.58.1
├─┬ @plotly/point-cluster@3.1.9
│ └── binary-search-bounds@2.0.4
├─┬ gl-contour2d@1.1.7
│ ├── binary-search-bounds@2.0.4
│ └─┬ cdt2d@1.0.0
│   └── binary-search-bounds@2.0.4
├─┬ gl-heatmap2d@1.1.0
│ └── binary-search-bounds@2.0.4
├─┬ gl-line3d@1.2.1
│ └── binary-search-bounds@2.0.4
├─┬ gl-plot2d@1.4.5
│ └── binary-search-bounds@2.0.4
├─┬ gl-plot3d@2.4.7
│ └─┬ 3d-view@2.0.0
│   └─┬ orbit-camera-controller@4.0.0
│     └─┬ filtered-vector@1.2.4
│       └── binary-search-bounds@1.0.0
├─┬ gl-scatter3d@1.2.3
│ └─┬ vectorize-text@3.2.1
│   └─┬ planar-graph-to-polyline@1.0.5
│     └─┬ point-in-big-polygon@2.0.0
│       ├── binary-search-bounds@1.0.0
│       ├─┬ interval-tree-1d@1.0.3
│       │ └── binary-search-bounds@1.0.0
│       └─┬ slab-decomposition@1.0.2
│         └── binary-search-bounds@1.0.0
├─┬ gl-surface3d@1.6.0
│ └── binary-search-bounds@2.0.4
└─┬ matrix-camera-controller@2.1.3
  └── binary-search-bounds@1.0.0
```

## at:
    file: binary-search-bounds/search-bounds.js
    func: compileBoundsSearch
    line: var result = new Function([

| bundle    | impacted |
| --------- | -------- |
| basic     | no       |
| cartesian | no       |
| finance   | no       |
| geo       | no       |
| mapbox    | no       |
| gl2d      | yes      |
| gl3d      | yes      |
---


# dependency: box-intersect
```
plotly.js@1.58.1
└─┬ gl-contour2d@1.1.7
  └─┬ clean-pslg@1.1.2
    └── box-intersect@1.0.2
```

## at:
    func: bruteForcePlanner
    file: box-intersect/lib/brute.js
    line: var proc = new Function(codeStr)

| bundle    | impacted |
| --------- | -------- |
| basic     | no       |
| cartesian | no       |
| finance   | no       |
| geo       | no       |
| mapbox    | no       |
| gl2d      | yes      |
| gl3d      | yes      |
---


# dependency: cwise-compiler
```
plotly.js@1.58.1
├─┬ gl-contour2d@1.1.7
│ └─┬ surface-nets@1.0.2
│   └─┬ zero-crossings@1.0.1
│     └── cwise-compiler@1.1.3
└─┬ gl-surface3d@1.6.0
  ├─┬ ndarray-gradient@1.0.0
  │ └── cwise-compiler@1.1.3
  ├─┬ ndarray-ops@1.2.2
  │ └── cwise-compiler@1.1.3
  └─┬ ndarray-pack@1.2.1
    └── cwise-compiler@1.1.3
```

## at:
    file: cwise-compiler/lib/compile.js
    func: generateCWiseOp
    line: var f = new Function(["function ",loopName,"(", arglist.join(","),"){", code.join("\n"),"} return ", loopName].join(""))

and
## at:
    func: createThunk
    file: cwise-compiler/lib/thunk.js
    line: var thunk = new Function("compile", code.join("\n"))

| bundle    | impacted |
| --------- | -------- |
| basic     | no       |
| cartesian | no       |
| finance   | no       |
| geo       | no       |
| mapbox    | no       |
| gl2d      | yes      |
| gl3d      | yes      |
---

# dependency: gl-shader
```
plotly.js@1.58.1
├─┬ gl-cone3d@1.5.2
│ └── gl-shader@4.2.1
├─┬ gl-contour2d@1.1.7
│ └── gl-shader@4.2.1
├─┬ gl-error3d@1.0.16
│ └── gl-shader@4.2.1
├─┬ gl-heatmap2d@1.1.0
│ └── gl-shader@4.2.1
├─┬ gl-line3d@1.2.1
│ └── gl-shader@4.2.1
├─┬ gl-mesh3d@2.3.1
│ └── gl-shader@4.2.1
├─┬ gl-plot2d@1.4.5
│ └── gl-shader@4.2.1
├─┬ gl-plot3d@2.4.7
│ ├─┬ gl-axes3d@1.5.3
│ │ └── gl-shader@4.2.1
│ ├── gl-shader@4.2.1
│ └─┬ gl-spikes3d@1.0.10
│   └── gl-shader@4.2.1
├─┬ gl-pointcloud2d@1.0.3
│ └── gl-shader@4.2.1
├─┬ gl-scatter3d@1.2.3
│ └── gl-shader@4.2.1
├─┬ gl-select-box@1.0.4
│ └── gl-shader@4.2.1
└─┬ gl-surface3d@1.6.0
  └── gl-shader@4.2.1
```

## at:
    file: gl-shader/lib/create-uniforms.js
    func: identity
    line: var c = new Function('y', 'return function(){return y}')

and
## at:
    file: gl-shader/lib/create-uniforms.js
    func: makeGetter
    line: var proc = new Function('gl', 'wrapper', 'locations', 'return function(){return gl.getUniform(wrapper.program,locations[' + index + '])}')

and
## at:
    file: gl-shader/lib/create-uniforms.js
    func: makeSetter
    line: var proc = new Function('gl', 'locations', code.join('\n'))

| bundle    | impacted |
| --------- | -------- |
| basic     | no       |
| cartesian | no       |
| finance   | no       |
| geo       | no       |
| mapbox    | no       |
| gl2d      | yes      |
| gl3d      | yes      |
---

# dependency: binary-search-bounds
```
plotly.js@1.58.1
├─┬ @plotly/point-cluster@3.1.9
│ └── binary-search-bounds@2.0.4
├─┬ gl-contour2d@1.1.7
│ ├── binary-search-bounds@2.0.4
│ └─┬ cdt2d@1.0.0
│   └── binary-search-bounds@2.0.4
├─┬ gl-heatmap2d@1.1.0
│ └── binary-search-bounds@2.0.4
├─┬ gl-line3d@1.2.1
│ └── binary-search-bounds@2.0.4
├─┬ gl-plot2d@1.4.5
│ └── binary-search-bounds@2.0.4
├─┬ gl-plot3d@2.4.7
│ └─┬ 3d-view@2.0.0
│   └─┬ orbit-camera-controller@4.0.0
│     └─┬ filtered-vector@1.2.4
│       └── binary-search-bounds@1.0.0
├─┬ gl-scatter3d@1.2.3
│ └─┬ vectorize-text@3.2.1
│   └─┬ planar-graph-to-polyline@1.0.5
│     └─┬ point-in-big-polygon@2.0.0
│       ├── binary-search-bounds@1.0.0
│       ├─┬ interval-tree-1d@1.0.3
│       │ └── binary-search-bounds@1.0.0
│       └─┬ slab-decomposition@1.0.2
│         └── binary-search-bounds@1.0.0
├─┬ gl-surface3d@1.6.0
│ └── binary-search-bounds@2.0.4
└─┬ matrix-camera-controller@2.1.3
  └── binary-search-bounds@1.0.0
```

## at:
    file: binary-search-bounds/search-bounds.js
    func: compileBoundsSearch
    line: var result = new Function([


| bundle    | impacted |
| --------- | -------- |
| basic     | no       |
| cartesian | no       |
| finance   | no       |
| geo       | no       |
| mapbox    | no       |
| gl2d      | yes      |
| gl3d      | yes      |
---

# dependency: ndarray-extract-contour
```
plotly.js@1.58.1
└─┬ gl-contour2d@1.1.7
  └─┬ surface-nets@1.0.2
    └── ndarray-extract-contour@1.0.1
```

## at:
    file: ndarray-extract-contour/contour.js
    func: compileSurfaceProcedure
    line: var proc = new Function("vertex", "face", "phase", "mallocUint32", "freeUint32", procedureCode)

| bundle    | impacted |
| --------- | -------- |
| basic     | no       |
| cartesian | no       |
| finance   | no       |
| geo       | no       |
| mapbox    | no       |
| gl2d      | yes      |
| gl3d      | yes      |
---


# dependency: ndarray-ops
```
plotly.js@1.58.1
├─┬ gl-cone3d@1.5.2
│ ├─┬ gl-buffer@2.1.2
│ │ └── ndarray-ops@1.2.2
│ └─┬ gl-texture2d@2.1.0
│   └── ndarray-ops@1.2.2
└─┬ gl-surface3d@1.6.0
  ├── ndarray-ops@1.2.2
  └─┬ ndarray-scratch@1.2.0
    └── ndarray-ops@1.2.2
```

## at:
    file: ndarray-ops/ndarray-ops.js
    func: makeOp
    line: var wrapper = new Function("P", [

| bundle    | impacted |
| --------- | -------- |
| basic     | no       |
| cartesian | no       |
| finance   | no       |
| geo       | no       |
| mapbox    | no       |
| gl2d      | yes      |
| gl3d      | yes      |
---


# dependency: ndarray
```
plotly.js@1.58.1
├─┬ gl-cone3d@1.5.2
│ ├─┬ gl-buffer@2.1.2
│ │ └── ndarray@1.0.19
│ ├─┬ gl-texture2d@2.1.0
│ │ └── ndarray@1.0.19
│ └── ndarray@1.0.19
├─┬ gl-contour2d@1.1.7
│ └── ndarray@1.0.19
├─┬ gl-line3d@1.2.1
│ └── ndarray@1.0.19
├─┬ gl-mesh3d@2.3.1
│ ├── ndarray@1.0.19
│ └─┬ simplicial-complex-contour@1.0.2
│   └── ndarray@1.0.19
├─┬ gl-plot2d@1.4.5
│ └─┬ gl-select-static@2.0.7
│   └── ndarray@1.0.19
├─┬ gl-plot3d@2.4.7
│ └── ndarray@1.0.19
├─┬ gl-scatter3d@1.2.3
│ └─┬ vectorize-text@3.2.1
│   └── ndarray@1.0.19
├─┬ gl-surface3d@1.6.0
│ ├── ndarray@1.0.19
│ ├─┬ ndarray-pack@1.2.1
│ │ └── ndarray@1.0.19
│ └─┬ ndarray-scratch@1.2.0
│   └── ndarray@1.0.19
└── ndarray@1.0.19
```

## at:
    file: ndarray/ndarray.js
    func: compileConstructor
    line: var procedure = new Function(code)

and
## at:
    file: ndarray/ndarray.js
    func: compileConstructor
    line: var procedure = new Function("TrivialArray", code)

and
## at:
    file: ndarray/ndarray.js
    func: compileConstructor
    line: var procedure = new Function("CTOR_LIST", "ORDER", code.join("\n"))

| bundle    | impacted |
| --------- | -------- |
| basic     | no       |
| cartesian | no       |
| finance   | no       |
| geo       | no       |
| mapbox    | no       |
| gl2d      | yes      |
| gl3d      | yes      |
---

# dependency: robust-in-sphere
```
plotly.js@1.58.1
└─┬ gl-contour2d@1.1.7
  └─┬ cdt2d@1.0.0
    └── robust-in-sphere@1.1.3
```

## at:
    file: robust-in-sphere/in-sphere.js
    line: var proc = new Function("sum", "diff", "prod", "scale", code.join(""))
    func: orientation

| bundle    | impacted |
| --------- | -------- |
| basic     | no       |
| cartesian | no       |
| finance   | no       |
| geo       | no       |
| mapbox    | no       |
| gl2d      | yes      |
| gl3d      | yes      |
---

# dependency: robust-orientation
```
plotly.js@1.58.1
├─┬ convex-hull@1.0.3
│ ├─┬ affine-hull@1.0.0
│ │ └── robust-orientation@1.1.3
│ ├─┬ incremental-convex-hull@1.0.1
│ │ └── robust-orientation@1.1.3
│ └─┬ monotone-convex-hull-2d@1.0.1
│   └── robust-orientation@1.1.3
├─┬ gl-contour2d@1.1.7
│ ├─┬ cdt2d@1.0.0
│ │ └── robust-orientation@1.1.3
│ └─┬ clean-pslg@1.1.2
│   └─┬ robust-segment-intersect@1.0.1
│     └── robust-orientation@1.1.3
├─┬ gl-plot3d@2.4.7
│ └─┬ gl-axes3d@1.5.3
│   └── robust-orientation@1.1.3
├─┬ gl-scatter3d@1.2.3
│ └─┬ vectorize-text@3.2.1
│   ├─┬ planar-graph-to-polyline@1.0.5
│   │ ├─┬ planar-dual@1.0.2
│   │ │ └─┬ compare-angle@1.0.1
│   │ │   └── robust-orientation@1.1.3
│   │ ├─┬ point-in-big-polygon@2.0.0
│   │ │ ├── robust-orientation@1.1.3
│   │ │ └─┬ slab-decomposition@1.0.2
│   │ │   └── robust-orientation@1.1.3
│   │ └── robust-orientation@1.1.3
│   └─┬ simplify-planar-graph@2.0.1
│     └── robust-orientation@1.1.3
└── robust-orientation@1.1.3
```

## at:
    file: robust-orientation/orientation.js
    func: orientation
    line: var proc = new Function("sum", "prod", "scale", "sub", code)

| bundle    | impacted |
| --------- | -------- |
| basic     | no       |
| cartesian | no       |
| finance   | no       |
| geo       | no       |
| mapbox    | no       |
| gl2d      | yes      |
| gl3d      | yes      |
---

# dependency: surface-nets
```
plotly.js@1.58.1
├─┬ gl-contour2d@1.1.7
│ └── surface-nets@1.0.2
├─┬ gl-scatter3d@1.2.3
│ └─┬ vectorize-text@3.2.1
│   └── surface-nets@1.0.2
└─┬ gl-surface3d@1.6.0
  └── surface-nets@1.0.2
```

## at:
    file: surface-nets/surfacenets.js
    func: buildSurfaceNets
    line: var proc = new Function("genContour", code.join(""))

| bundle    | impacted |
| --------- | -------- |
| basic     | no       |
| cartesian | no       |
| finance   | no       |
| geo       | no       |
| mapbox    | no       |
| gl2d      | yes      |
| gl3d      | yes      |
---


# dependency: incremental-convex-hull
```
plotly.js@1.58.1
├─┬ convex-hull@1.0.3
│ └── incremental-convex-hull@1.0.1
└─┬ delaunay-triangulate@1.1.6
  └── incremental-convex-hull@1.0.1
```

## at:
    file: incremental-convex-hull/ich.js
    func: bakeOrient
    line: var proc = new Function("test", code.join(""))

| bundle    | impacted |
| --------- | -------- |
| basic     | no       |
| cartesian | no       |
| finance   | no       |
| geo       | no       |
| mapbox    | no       |
| gl2d      | no       |
| gl3d      | yes      |
---


# dependency: ndarray-sort
```
plotly.js@1.58.1
└─┬ gl-mesh3d@2.3.1
  └─┬ simplicial-complex-contour@1.0.2
    └── ndarray-sort@1.0.1
```

## at:
    file: ndarray-sort/lib/compile_sort.js
    func: createInsertionSort
    line: var result = new Function("malloc", "free", code.join("\n"))

and
## at:
    file: ndarray-sort/lib/compile_sort.js
    func: createInsertionSort
    line: var result = new Function(code.join("\n"))

and
## at:
    file: ndarray-sort/lib/compile_sort.js
    func: createQuickSort
    line: var compiled = new Function("insertionSort", "malloc", "free", code.join("\n"))

and
## at:
    file: ndarray-sort/lib/compile_sort.js
    func: createQuickSort
    line: var new Function("insertionSort", code.join("\n"))

and
## at:
    file: ndarray-sort/lib/compile_sort.js
    func: createQuickSort
    line: var result = new Function("insertionSort", "quickSort", code.join("\n"))

| bundle    | impacted |
| --------- | -------- |
| basic     | no       |
| cartesian | no       |
| finance   | no       |
| geo       | no       |
| mapbox    | no       |
| gl2d      | no       |
| gl3d      | yes      |
---


# dependency: robust-determinant
```
plotly.js@1.58.1
└─┬ gl-mesh3d@2.3.1
  └─┬ barycentric@1.0.1
    └─┬ robust-linear-solve@1.0.0
      └── robust-determinant@1.1.0
```

## at:
    file: robust-determinant/robust-determinant.js
    func: compileDeterminant
    line: var proc = new Function("sum", "scale", "prod", "compress", [

| bundle    | impacted |
| --------- | -------- |
| basic     | no       |
| cartesian | no       |
| finance   | no       |
| geo       | no       |
| mapbox    | no       |
| gl2d      | no       |
| gl3d      | yes      |
---

# dependency: robust-linear-solve
```
plotly.js@1.58.1
├─┬ alpha-shape@1.0.0
│ └─┬ alpha-complex@1.0.0
│   └─┬ circumradius@1.0.0
│     └─┬ circumcenter@1.0.0
│       └── robust-linear-solve@1.0.0
└─┬ gl-mesh3d@2.3.1
  └─┬ barycentric@1.0.1
    └── robust-linear-solve@1.0.0
```

## at:
    file: robust-linear-solve/linsolve.js
    func: generateSolver
    line: var proc = new Function("det", code.join(""))

| bundle    | impacted |
| --------- | -------- |
| basic     | no       |
| cartesian | no       |
| finance   | no       |
| geo       | no       |
| mapbox    | no       |
| gl2d      | no       |
| gl3d      | yes      |
---

# dependency: simplicial-complex-contour
```
plotly.js@1.58.1
└─┬ gl-mesh3d@2.3.1
  └── simplicial-complex-contour@1.0.2
```

## at:
    file: simplicial-complex-contour/lib/codegen.js
    func: createCellPolygonizer
    line: var proc = new Function('pool', code.join(''))

| bundle    | impacted |
| --------- | -------- |
| basic     | no       |
| cartesian | no       |
| finance   | no       |
| geo       | no       |
| mapbox    | no       |
| gl2d      | no       |
| gl3d      | yes      |
---

**Comment by tdelmas** [comment]
@archmoj Here is the list of GitHub repository involved in the list you provided, with, when available, the issue and/or the PR about CSP:

- d3 (d3/src/dsv/dsv.js): https://github.com/d3/d3-dsv/issues/65 / https://github.com/d3/d3-fetch/issues/35
- es6-promise: shouldn't be a problem, the line https://github.com/stefanpenner/es6-promise/blame/f97e2666e6928745c450752e74213d2438b48b4c/lib/es6-promise/asap.js#L99 doesn't seem to be reachable using a modern Browser
- binary-search-bounds by @mikolalysenko: https://github.com/mikolalysenko/binary-search-bounds/issues/5 and https://github.com/mikolalysenko/binary-search-bounds/pull/8
- box-intersect by @mikolalysenko: https://github.com/mikolalysenko/box-intersect/issues/6
- https://github.com/scijs/cwise-compiler by @mikolalysenko
- https://github.com/stackgl/gl-shader by @mikolalysenko: https://github.com/stackgl/gl-shader/issues/21
- https://github.com/scijs/ndarray-extract-contour by @mikolalysenko
- https://github.com/scijs/ndarray-ops by @mikolalysenko
- https://github.com/scijs/ndarray by @mikolalysenko
- https://github.com/mikolalysenko/robust-in-sphere by @mikolalysenko
- https://github.com/mikolalysenko/robust-orientation by @mikolalysenko
- https://github.com/mikolalysenko/surface-nets by @mikolalysenko
- https://github.com/mikolalysenko/incremental-convex-hull by @mikolalysenko
- https://github.com/scijs/ndarray-sort by @mikolalysenko
- https://github.com/mikolalysenko/robust-determinant by @mikolalysenko
- https://github.com/mikolalysenko/robust-linear-solve by @mikolalysenko
- https://github.com/mikolalysenko/simplicial-complex-contour by @mikolalysenko

As @mikolalysenko / @scijs doesn't seam to want to address the problem, most of the dependency you cited (probably around 50 packages?) may require to be forked...

**Comment by nschloe** [comment]
Big thanks for releasing [plotly-strict](https://www.npmjs.com/package/plotly.js-strict-dist)! This actually made it possible for me to use it in [a browser extension](https://chrome.google.com/webstore/detail/xdoc/anidddebgkllnnnnjfkmjcaallemhjee) for displaying plotly graphs in GitHub READMEs etc.:

![screenshot](https://user-images.githubusercontent.com/181628/122204266-389f0680-ce9f-11eb-9a1f-ec23d8b22092.png)

This wasn't possible before because browser extensions disallow `eval()`/function constructors.

**Comment by archmoj** [comment]
The number of function constructors were slightly higher that those initially listed above. 
Namely `Function.apply` calls were not catch by `eslint` test. 
The list is now updated on #5869.

**Comment by archmoj** [comment]
:tada: The [v2.5.0](https://github.com/plotly/plotly.js/releases/tag/v2.5.0) release removed calls to function constructor from various dependencies of `gl3d` traces as well as `heatmapgl` and `pointcloud`.

It is now possible to apply CSP directives e.g. https://github.com/plotly/plotly.js/blob/8788e57b7597ca1427d508fac8d26967781b230e/devtools/test_dashboard/index-strict.html#L4 to render & interact with those `stackgl-based` traces using `"strict"` or `"gl3d"` bundles.
